File: Android/Sources/UserControlLibraries/HMSInAppBillingLib/src/main/java/com/genexus/android/payments/hms/StoreManager.java
package com.genexus.android.payments.hms;

import android.app.Activity;
import android.content.Intent;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.genexus.android.core.actions.ApiAction;
import com.genexus.android.core.activities.ActivityHelper;
import com.genexus.android.StoreManagerApi;
import com.genexus.android.core.base.model.Entity;
import com.genexus.android.core.base.model.ValueCollection;
import com.genexus.android.core.base.services.Services;
import com.genexus.android.core.externalapi.ExternalApiResult;
import com.genexus.android.payments.hms.common.CipherUtil;
import com.genexus.android.payments.hms.common.Constants;
import com.genexus.android.payments.hms.common.IapRequestHelper;
import com.huawei.hms.api.ConnectionResult;
import com.huawei.hms.api.HuaweiApiAvailability;
import com.huawei.hms.iap.Iap;
import com.huawei.hms.iap.IapClient;
import com.huawei.hms.iap.entity.ConsumeOwnedPurchaseResult;
import com.huawei.hms.iap.entity.InAppPurchaseData;
import com.huawei.hms.iap.entity.IsEnvReadyResult;
import com.huawei.hms.iap.entity.OrderStatusCode;
import com.huawei.hms.iap.entity.OwnedPurchasesResult;
import com.huawei.hms.iap.entity.ProductInfo;
import com.huawei.hms.iap.entity.ProductInfoResult;
import com.huawei.hms.iap.entity.PurchaseIntentResult;
import com.huawei.hms.iap.entity.PurchaseResultInfo;

import org.json.JSONException;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class StoreManager extends StoreManagerApi {
	public static final String OBJECT_NAME = "GeneXus.SD.Store.StoreManager";

	private static final String READ_PROPERTY_CAN_MAKE_PURCHASES = "CanMakePurchases";
	private static final String METHOD_GET_PRODUCTS = "GetProducts";
	private static final String METHOD_PURCHASE_PRODUCT = "purchaseProduct";
	private static final String METHOD_CONSUME_PRODUCT = "ConsumeProduct";
	private static final String METHOD_GET_PURCHASES = "GetPurchases";
	@SuppressWarnings("unused")
	private static final String METHOD_IS_PRODUCT_ENABLED = "IsEnabled";
	@SuppressWarnings("unused")
	private static final String METHOD_ENABLE_PRODUCT = "EnableProduct";
	@SuppressWarnings("unused")
	private static final String METHOD_DISABLE_PRODUCT = "DisableProduct";

	private IapClient mIapClient = null;
	private static final String SUBSCRIPTIONNAME = "subscription";
	private static final String SUSCRIPTIONNAME = "suscription";

	private Entity mPurchaseResult;

	public StoreManager(ApiAction action) {
		super(action);
		mIapClient = Iap.getIapClient(this.getActivity());
		addSimpleMethodHandler(READ_PROPERTY_CAN_MAKE_PURCHASES, 0, mPropertyCanMakePurchases);
		addSimpleMethodHandler(METHOD_GET_PRODUCTS, 1, mMethodGetProducts);
		addMethodHandler(METHOD_PURCHASE_PRODUCT, 2, mMethodPurchaseProduct);
		addMethodHandler(METHOD_PURCHASE_PRODUCT, 1, mMethodPurchaseProduct);
		addSimpleMethodHandler(METHOD_GET_PURCHASES, 0, mMethodGetPurchases);
		addSimpleMethodHandler(METHOD_CONSUME_PRODUCT, 1, mMethodConsumeProduct);
	}

	private final ISimpleMethodInvoker mPropertyCanMakePurchases = new ISimpleMethodInvoker() {
		@Override
		public Object invoke(List<Object> parameters) {
			int result = HuaweiApiAvailability.getInstance().isHuaweiMobileServicesAvailable(getContext());
			if (result == ConnectionResult.SUCCESS)
			{
				IsEnvReadyResult envReadyResult = IapRequestHelper.isEnvReadyNow(mIapClient);
				if (envReadyResult!=null)
				{
					return envReadyResult.getReturnCode()==ConnectionResult.SUCCESS;
				}
			}
			return false;
		}
	};

	private final ISimpleMethodInvoker mMethodGetProducts = new ISimpleMethodInvoker() {
		@Override
		public Object invoke(List<Object> parameters) {
			ValueCollection skus = (ValueCollection) parameters.get(0);
			List<ProductDetail> productDetails =null;

			// assume default type as CONSUMABLE
			int type = guessProductsType(skus);
			productDetails = new ArrayList<>();

			ProductInfoResult productInfoResult = IapRequestHelper.obtainProductInfoNow(mIapClient, skus, type );
			OwnedPurchasesResult ownedPurchasesResult = IapRequestHelper.obtainOwnedPurchasesNow(mIapClient, type, null);

			if (productInfoResult!=null && ownedPurchasesResult!=null) {
				List<String> inAppPurchaseDataList = ownedPurchasesResult.getInAppPurchaseDataList();
				List<ProductInfo> productInfos = productInfoResult.getProductInfoList();
				for (ProductInfo productInfo : productInfos) {
					// check if product is purchased
					boolean purchased = false;
					for (String data : inAppPurchaseDataList) {
						try {
							InAppPurchaseData inAppPurchaseData = new InAppPurchaseData(data);
							if (productInfo.getProductId().equals(inAppPurchaseData.getProductId())) {
								purchased = true;
							}
						} catch (JSONException e) {
							Services.Log.error("parse InAppPurchaseData JSONException", e);
						}
					}

					ProductDetail productDetail = new ProductDetail(
						productInfo.getProductId(),
						productInfo.getProductName(),
						productInfo.getProductDesc(),
						productInfo.getPrice(),
						purchased
					);
					productDetails.add(productDetail);
				}
			}
			else
			{
				Services.Log.error("MethodGetProducts, error getting products list");
			}

			return StoreEntitiesFactory.createStoreProductCollection(productDetails);
		}
	};


	private final IMethodInvoker mMethodPurchaseProduct = new IMethodInvokerWithActivityResult() {
		@Override
		public @NonNull ExternalApiResult invoke(List<Object> parameters) {
			String sku = (String) parameters.get(0);

			if (parameters.size() > 1) {
				int productQty = Integer.parseInt(((String) parameters.get(1)));

				// HMS's In-app Billing API only supports buying one product at a time.
				if (productQty != 1) {
					mPurchaseResult = StoreEntitiesFactory.createEmptyPurchaseResult();
					return ExternalApiResult.success(mPurchaseResult);
				}
			}

			mPurchaseResult = null;
			List<String> skus = Collections.singletonList(sku);

			// assume default type as CONSUMABLE
			int type = guessProductsType(skus);

			PurchaseIntentResult purchaseIntentResult = IapRequestHelper.createPurchaseIntentNow(mIapClient, sku, type );

			if (purchaseIntentResult!=null)
			{
				// you should pull up the page to complete the payment process
				ActivityHelper.registerActionRequestCode(Constants.REQ_CODE_BUY);
				IapRequestHelper.startResolutionForResult( getActivity(), purchaseIntentResult.getStatus(), Constants.REQ_CODE_BUY);
				return ExternalApiResult.SUCCESS_WAIT;
			}
			else
			{
				Services.Log.error("GetBuyIntentResult is null");
				mPurchaseResult = StoreEntitiesFactory.createEmptyPurchaseResult();
				return ExternalApiResult.success(mPurchaseResult);
			}


		}

		@Override
		public @NonNull ExternalApiResult handleActivityResult(int requestCode, int resultCode, @Nullable Intent result) {
			Services.Log.debug("PurchaseProduct, handleActivityResult");
			if (requestCode == Constants.REQ_CODE_BUY) {
				if (resultCode == Activity.RESULT_OK) {
					Services.Log.debug("PurchaseProduct, handleActivityResult RESULT_OK");
					PurchaseResultInfo purchaseIntentResult = mIapClient.parsePurchaseResultInfoFromIntent(result);
					switch(purchaseIntentResult.getReturnCode()) {
						case OrderStatusCode.ORDER_STATE_CANCEL:
							//Toast.makeText(this, R.string.ordercancel, Toast.LENGTH_SHORT).show();
							Services.Log.info("PurchaseProduct, ORDER_STATE_CANCEL purchase");
							break;
						case OrderStatusCode.ORDER_STATE_FAILED:
						case OrderStatusCode.ORDER_PRODUCT_OWNED:
							Services.Log.info("PurchaseProduct, ORDER_STATE_FAILED or  ORDER_PRODUCT_OWNED");
							break;
						case OrderStatusCode.ORDER_STATE_SUCCESS:
							Services.Log.info("PurchaseProduct, ORDER_STATE_SUCCESS");
							boolean credible = CipherUtil.doCheck(purchaseIntentResult.getInAppPurchaseData(), purchaseIntentResult.getInAppDataSignature(), CipherUtil.getPublicKey());
							Services.Log.info("PurchaseProduct, doCheck result " + credible);
							if (credible) {
								try {
									InAppPurchaseData inAppPurchaseData = new InAppPurchaseData(purchaseIntentResult.getInAppPurchaseData());

									// check product type to validate
									String sku = inAppPurchaseData.getProductId();
									List<String> skus = Collections.singletonList(sku);
									// assume default type as CONSUMABLE
									int type = guessProductsType(skus);

									// validate subscription
									if (type == IapClient.PriceType.IN_APP_SUBSCRIPTION)
									{
										if (inAppPurchaseData.isSubValid()) {
											Services.Log.info("PurchaseProduct, ORDER_STATE_SUCCESS and VALID SUBVALID");
											mPurchaseResult = StoreEntitiesFactory.createSuccessfulPurchaseResult(inAppPurchaseData.getProductId(), purchaseIntentResult.getInAppPurchaseData(), inAppPurchaseData.getOrderID());
											return ExternalApiResult.successNoRefresh(mPurchaseResult);
										}
										else
											Services.Log.error("PurchaseProduct, inAppPurchaseData isSubValid NOT CORRECT");
									}
									else // validate consumable
									{
										if (inAppPurchaseData.getPurchaseState() == InAppPurchaseData.PurchaseState.PURCHASED)
										{
											Services.Log.info("PurchaseProduct, ORDER_STATE_SUCCESS and VALID PURCHASED");
											mPurchaseResult = StoreEntitiesFactory.createSuccessfulPurchaseResult(inAppPurchaseData.getProductId(), purchaseIntentResult.getInAppPurchaseData(), inAppPurchaseData.getOrderID());
											return ExternalApiResult.successNoRefresh(mPurchaseResult);
										}
										else
											Services.Log.error("PurchaseProduct, inAppPurchaseData getPurchaseState NOT PURCHASED");
									}
								} catch (JSONException e) {
									Services.Log.error("PurchaseProduct, parse InAppPurchaseData JSONException", e);
								}
							} else {
								Services.Log.error("PurchaseProduct, check the data signature fail");
							}
							break;
						default:
							break;
					}
				} else {
					Services.Log.info("PurchaseProduct, cancel purchase");
				}
			}
			mPurchaseResult = StoreEntitiesFactory.createEmptyPurchaseResult();
			return ExternalApiResult.successNoRefresh(mPurchaseResult);
		}

	};

	private ISimpleMethodInvoker mMethodGetPurchases = new ISimpleMethodInvoker() {
		@Override
		public Object invoke(List<Object> parameters) {
			OwnedPurchasesResult ownedPurchasesResult = IapRequestHelper.obtainOwnedPurchasesNow(mIapClient, IapClient.PriceType.IN_APP_CONSUMABLE, null);

			Entity purchasesInfo = StoreEntitiesFactory.createEmptyPurchasesInformation();
			if (ownedPurchasesResult!=null) {
				purchasesInfo = StoreEntitiesFactory.createPurchasesInformation(ownedPurchasesResult);
			}

			OwnedPurchasesResult ownedPurchasesResult2 = IapRequestHelper.obtainOwnedPurchasesNow(mIapClient, IapClient.PriceType.IN_APP_SUBSCRIPTION, null);
			if (ownedPurchasesResult!=null) {
				if (purchasesInfo==null)
					purchasesInfo = StoreEntitiesFactory.createPurchasesInformation(ownedPurchasesResult2);
				else
					purchasesInfo = StoreEntitiesFactory.addPurchasesInformation(ownedPurchasesResult2, purchasesInfo);
			}
			return purchasesInfo;

		}
	};

	private ISimpleMethodInvoker mMethodConsumeProduct = new ISimpleMethodInvoker() {
		@Override
		public Object invoke(List<Object> parameters) {
			String sku = (String) parameters.get(0);
			boolean success = false;

			// only consumable can be consume
			OwnedPurchasesResult ownedPurchasesResult = IapRequestHelper.obtainOwnedPurchasesNow(mIapClient, IapClient.PriceType.IN_APP_CONSUMABLE, null);
			if (ownedPurchasesResult!=null) {
				List<String> inAppPurchaseDataList = ownedPurchasesResult.getInAppPurchaseDataList();

				for (String data : inAppPurchaseDataList) {
					try {
						InAppPurchaseData inAppPurchaseData = new InAppPurchaseData(data);
						if (inAppPurchaseData.getProductId().equalsIgnoreCase(sku)) {
							String purchaseToken = inAppPurchaseData.getPurchaseToken();
							ConsumeOwnedPurchaseResult result = IapRequestHelper.consumeOwnedPurchaseNow(mIapClient, purchaseToken);
							if (result != null)
								success = true;
						}
					} catch (JSONException e) {
						Services.Log.error("parse InAppPurchaseData JSONException", e);
					}
				}
			}
			return success;
		}
	};

	private int guessProductsType(List<String> skus) {
		int type = IapClient.PriceType.IN_APP_CONSUMABLE;
		for (String sku : skus) {
			// use SUBSCRIPTION, if product name contains it.
			if (sku.toLowerCase().contains(SUBSCRIPTIONNAME) || sku.toLowerCase().contains(SUSCRIPTIONNAME))
				type = IapClient.PriceType.IN_APP_SUBSCRIPTION;
		}
		return type;
	}


}

File: Android/Sources/UserControlLibraries/InAppBillingLib/src/main/java/com/genexus/android/payments/google/StoreManager.kt
package com.genexus.android.payments.google

import androidx.lifecycle.Lifecycle
import androidx.lifecycle.LifecycleObserver
import androidx.lifecycle.LifecycleOwner
import androidx.lifecycle.OnLifecycleEvent
import com.android.billingclient.api.BillingResult
import com.android.billingclient.api.Purchase
import com.genexus.android.StoreManagerApi
import com.genexus.android.core.actions.ActionExecution
import com.genexus.android.core.actions.ApiAction
import com.genexus.android.core.base.metadata.expressions.Expression
import com.genexus.android.core.base.model.Entity
import com.genexus.android.core.base.model.EntityFactory
import com.genexus.android.core.base.model.ValueCollection
import com.genexus.android.core.base.services.Services
import com.genexus.android.core.externalapi.ExternalApi.IMethodInvoker
import com.genexus.android.core.externalapi.ExternalApiResult
import com.genexus.android.payments.google.StoreEntitiesFactory.createEmptyPurchaseResult
import com.genexus.android.payments.google.StoreEntitiesFactory.createFailedPurchaseResult
import com.genexus.android.payments.google.StoreEntitiesFactory.createPurchasesInformation
import com.genexus.android.payments.google.StoreEntitiesFactory.createStoreProductCollection
import com.genexus.android.payments.google.StoreEntitiesFactory.createSuccessfulPurchaseResult
import com.genexus.android.payments.google.util.BillingUpdatedListener
import com.genexus.android.payments.google.util.ProductDetail
import com.google.android.gms.common.ConnectionResult
import com.google.android.gms.common.GoogleApiAvailability
import java.util.concurrent.atomic.AtomicBoolean

class StoreManager(action: ApiAction?) : StoreManagerApi(action), LifecycleObserver {
    private val operationInProgress = AtomicBoolean(false)
    private val disposeRequested = AtomicBoolean(false)
    private var purchaseResult: Entity? = null
    private val purchaseUpdatedListener: BillingUpdatedListener = object : BillingUpdatedListener {
        override fun onPurchaseUpdated(result: BillingResult, purchase: Purchase) {
            purchaseResult = if (service.isBillingOk(result))
                createSuccessfulPurchaseResult(purchase.skus[0], purchase.originalJson, purchase.purchaseToken)
            else
                createFailedPurchaseResult(result.debugMessage)

            onOperationEnded()
            action?.let {
                it.setOutputValue(Expression.Value.newSdt(purchaseResult))
                ActionExecution.continueCurrent(activity, true, it)
            }
        }
    }

    private val service = IabService(activity.applicationContext, purchaseUpdatedListener)
    private var lifecycle: Lifecycle? = null

    private val propertyCanMakePurchases = IMethodInvoker {
        val result = GoogleApiAvailability.getInstance().isGooglePlayServicesAvailable(context) == ConnectionResult.SUCCESS && service.isAvailable()
        ExternalApiResult.success(result)
    }

    private val methodGetProducts = IMethodInvoker { parameters: List<Any?> ->
        val skus = parameters[0] as ValueCollection
        var productDetails: List<ProductDetail>
        try {
            onOperationStarted()
            val inventory = service.getSkuDetails(skus)
            productDetails = ArrayList()
            for (details in inventory) {
                val productDetail = ProductDetail(
                    details.sku, details.title,
                    details.description, details.price, service.hasPurchase(details.sku)
                )

                productDetails.add(productDetail)
            }
        } catch (e: IllegalStateException) {
            productDetails = emptyList()
        } finally {
            onOperationEnded()
        }
        ExternalApiResult.success(createStoreProductCollection(productDetails))
    }

    private val methodPurchaseProduct = IMethodInvoker { parameters ->
        val sku = parameters[0] as String
        if (parameters.size > 1) {
            val productQty = (parameters[1] as String).toInt()

            // Google's In-app Billing API only supports buying one product at a time.
            if (productQty != 1)
                return@IMethodInvoker ExternalApiResult.success(createEmptyPurchaseResult())
        }

        /* In case we have the sku registered as purchased in our inventory, the onPurchaseFinishedListener
			is called directly by launchPurchaseFlow. So after it finishes we already have our PurchaseResult
			ready and want to return the result immediately.
			Otherwise, onPurchaseFinishedListener is called by handleActivityResult which should be called on
			ActivityResult, so we want to defer returning our PurchaseResult to afterActivityResult.
		 */
        try {
            onOperationStarted()
            if (!service.launchPurchaseFlow(activity, sku))
                throw IllegalStateException("BillingClient is null")

            return@IMethodInvoker ExternalApiResult.SUCCESS_WAIT
        } catch (e: IllegalStateException) {
            Services.Log.debug(OBJECT_NAME, String.format("PurchaseProduct failed. Reason: %s", e.message))
            return@IMethodInvoker ExternalApiResult.success(createEmptyPurchaseResult())
        } finally {
            onOperationEnded()
        }
    }

    private val methodGetPurchases = IMethodInvoker {
        var purchases: List<Purchase>
        try {
            onOperationStarted()
            purchases = service.getPurchases()
        } catch (e: IllegalStateException) {
            purchases = emptyList()
            Services.Log.debug(OBJECT_NAME, String.format("GetPurchases failed. Reason: %s", e.message))
        } finally {
            onOperationEnded()
        }
        ExternalApiResult.success(createPurchasesInformation(purchases))
    }

    private val methodConsumeProduct = IMethodInvoker { parameters: List<Any?> ->
        val sku = parameters[0] as String
        val success: Boolean = try {
            onOperationStarted()
            if (!service.hasPurchase(sku))
                false
            else {
                val purchases = service.getPurchases(sku)
                if (purchases.isEmpty())
                    false
                else
                    service.consumeOrAcknowledgePurchase(purchases[0])
            }
        } catch (e: IllegalStateException) {
            Services.Log.debug(OBJECT_NAME, String.format("Consume failed. Reason: %s", e.message))
            false
        } finally {
            onOperationEnded()
        }
        ExternalApiResult.success(success)
    }

    private val methodRestorePurchases = IMethodInvoker {
        ExternalApiResult.success(EntityFactory.newEntity())
    }

    private fun onOperationStarted() {
        operationInProgress.compareAndSet(false, true)
    }

    private fun onOperationEnded() {
        operationInProgress.compareAndSet(true, false)
        if (disposeRequested.get()) {
            service.dispose()
            disposeRequested.compareAndSet(true, false)
        }
    }

    @OnLifecycleEvent(Lifecycle.Event.ON_DESTROY)
    private fun onActivityDestroyed() {
        if (operationInProgress.get())
            disposeRequested.compareAndSet(false, true)
        else
            service.dispose()

        lifecycle?.let { Services.Device.runOnUiThread { it.removeObserver(this) } }
    }

    companion object {
        const val OBJECT_NAME = "GeneXus.SD.Store.StoreManager"
        private const val READ_PROPERTY_CAN_MAKE_PURCHASES = "CanMakePurchases"
        private const val METHOD_GET_PRODUCTS = "GetProducts"
        private const val METHOD_PURCHASE_PRODUCT = "PurchaseProduct"
        private const val METHOD_CONSUME_PRODUCT = "ConsumeProduct"
        private const val METHOD_GET_PURCHASES = "GetPurchases"
        private const val METHOD_RESTORE_PURCHASES = "RestorePurchases"
    }

    init {
        addMethodHandler(READ_PROPERTY_CAN_MAKE_PURCHASES, 0, propertyCanMakePurchases)
        addMethodHandler(METHOD_GET_PRODUCTS, 1, methodGetProducts)
        addMethodHandler(METHOD_PURCHASE_PRODUCT, 2, methodPurchaseProduct)
        addMethodHandler(METHOD_PURCHASE_PRODUCT, 1, methodPurchaseProduct)
        addMethodHandler(METHOD_GET_PURCHASES, 0, methodGetPurchases)
        addMethodHandler(METHOD_CONSUME_PRODUCT, 1, methodConsumeProduct)
        addMethodHandler(METHOD_RESTORE_PURCHASES, 0, methodRestorePurchases)

        (activity as? LifecycleOwner)?.lifecycle?.let {
            Services.Device.runOnUiThread { it.addObserver(this) }
            lifecycle = it
        }
    }
}
